contrib/artifactory/src/mill/contrib/artifactory/ArtifactoryPublishModule.scala

```diff
@@ -51,6 +73,23 @@ object ArtifactoryPublishModule extends ExternalModule {
     )
   }
 
+  private def checkArtifactoryCreds(credentials: String)(implicit ctx: mill.api.Ctx.Env): String = {
+    if (credentials.isEmpty) {
+      (for {
+        username <- T.env.get("ARTIFACTORY_USERNAME")
+        password <- T.env.get("ARTIFACTORY_PASSWORD")
+      } yield {
+        s"$username:$password"
+      }).getOrElse(
+        throw new IllegalArgumentException(
+          "Consider using ARTIFACTORY_USERNAME/ARTIFACTORY_PASSWORD environment variables or passing `credentials` argument"
+        )
+      )
+    } else {
+      credentials
+    }
+  }
+
```

Comment:
If you make this `def` return a `Task[String]` with `T.task { }`, you'd get a proper way to return a `Result.Failure` which can then nicely bubble up the task graph and reported by Mill. That way, you also don't need to declare the implicit `ctx`, as the task already has it.